homeassistant/components/volvooncall.py

```diff
@@ -85,52 +70,76 @@ def setup(hass, config):
 
     def discover_vehicle(vehicle):
         """Load relevant platforms."""
-        state.entities[vehicle.vin] = []
-        for attr, (component, *_) in RESOURCES.items():
-            if (getattr(vehicle, attr + '_supported', True) and
-                    attr in config[DOMAIN].get(CONF_RESOURCES, [attr])):
-                discovery.load_platform(
-                    hass, component, DOMAIN, (vehicle.vin, attr), config)
+        state.vehicles.append(vehicle.vin)
 
-    def update_vehicle(vehicle):
-        """Receive updated information on vehicle."""
-        state.vehicles[vehicle.vin] = vehicle
-        if vehicle.vin not in state.entities:
-            discover_vehicle(vehicle)
+        dashboard = vehicle.dashboard(
+            mutable=config[DOMAIN].get(CONF_MUTABLE),
+            scandinavian_miles=config[DOMAIN].get(CONF_SCANDINAVIAN_MILES))
 
-        for entity in state.entities[vehicle.vin]:
-            entity.schedule_update_ha_state()
+        def is_enabled(attr):
+            return attr in config[DOMAIN].get(CONF_RESOURCES, [attr])
 
-        dispatcher_send(hass, SIGNAL_VEHICLE_SEEN, vehicle)
+        for instrument in (
+                instrument
+                for instrument in dashboard.instruments
+                if is_enabled(instrument.slug_attr)):
 
-    def update(now):
+            state.instruments.append(instrument)
+
+            hass.async_create_task(
+                discovery.async_load_platform(
+                    hass,
+                    instrument.component,
```

Comment:
We should decouple `instrument.component` which is part of volvooncall api from the home assistant component name which is part of the home assistant api. Use a dict to lookup valid home assistant components before loading platforms.

Reply:
I can add that. The component names in the dependency are supposed to mirror the ones in hass 1:1, since the library also works as an MQTT gateway (this is how I use it), for which the hass component names are needed anyway (for MQTT discovery).

Reply:
Libraries should preferably not be aware of home assistant, and we should never depend on that in home assistant.

Reply:
✓